src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -345,15 +395,34 @@ public void OnResponseAbort(Exception abortException)
                 }
             }
 
-            public async Task ReadResponseHeadersAsync()
+            public async Task ReadResponseHeadersAsync(TaskCompletionSource<bool> tcs = null)
             {
                 // Wait for response headers to be read.
-                (bool wait, bool emptyResponse) = TryEnsureHeaders();
-                if (wait)
+                bool emptyResponse;
+                bool wait;
+
+                do
                 {
-                    await GetWaiterTask().ConfigureAwait(false);
                     (wait, emptyResponse) = TryEnsureHeaders();
-                    Debug.Assert(!wait);
+                    if (wait)
+                    {
+                        await GetWaiterTask().ConfigureAwait(false);
+                        (wait, emptyResponse) = TryEnsureHeaders();
+                        Debug.Assert(!wait);
+                    }
+
+                    if (Response.StatusCode  == HttpStatusCode.Continue)
+                    {
+                        OnResponse100Continue();
+                        tcs?.TrySetResult(true);
```

Comment:
Seems like we should set tcs to null here, so we don't try to complete it twice it the unlikely event that we receive more than one 100 continue response.